File: examples/ble/android/app/src/main/java/dev/ndnrs/bletest/BleClient.kt
package dev.ndnrs.bletest

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.os.ParcelUuid
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow

/**
 * BLE GATT central client for NDN-over-BLE testing.
 *
 * Scans for the NDN BLE service, connects, and provides read/write access
 * to the CS (write) and SC (notify) characteristics.
 */
@SuppressLint("MissingPermission")
class BleClient(private val context: Context) {

    sealed class State {
        data object Idle : State()
        data object Scanning : State()
        data class Found(val device: BluetoothDevice, val name: String?) : State()
        data object Connecting : State()
        data object Connected : State()
        data object Disconnected : State()
    }

    data class ScannedDevice(val device: BluetoothDevice, val name: String?)

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state

    private val _events = MutableSharedFlow<String>(extraBufferCapacity = 64)
    val events: SharedFlow<String> = _events

    private val _rxData = MutableSharedFlow<ByteArray>(extraBufferCapacity = 16)
    val rxData: SharedFlow<ByteArray> = _rxData

    private val _scannedDevices = MutableStateFlow<List<ScannedDevice>>(emptyList())
    val scannedDevices: StateFlow<List<ScannedDevice>> = _scannedDevices

    private val adapter: BluetoothAdapter? =
        (context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter

    private var scanner: BluetoothLeScanner? = null
    private var gatt: BluetoothGatt? = null
    private var csChar: BluetoothGattCharacteristic? = null
    private var scChar: BluetoothGattCharacteristic? = null

    // ── Scanning ─────────────────────────────────────────────────────────

    fun startScan() {
        val adapter = this.adapter ?: run {
            _events.tryEmit("Bluetooth not available")
            return
        }
        if (!adapter.isEnabled) {
            _events.tryEmit("Bluetooth is disabled")
            return
        }

        scanner = adapter.bluetoothLeScanner
        _scannedDevices.value = emptyList()
        _state.value = State.Scanning

        val filter = ScanFilter.Builder()
            .setServiceUuid(ParcelUuid(NdnBleProtocol.SERVICE_UUID))
            .build()
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        _events.tryEmit("Scanning for NDN BLE peripherals …")
        scanner?.startScan(listOf(filter), settings, scanCallback)
    }

    fun stopScan() {
        scanner?.stopScan(scanCallback)
        if (_state.value == State.Scanning) {
            _state.value = State.Idle
        }
    }

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val device = result.device
            val name = result.scanRecord?.deviceName ?: device.name
            val current = _scannedDevices.value
            if (current.none { it.device.address == device.address }) {
                _scannedDevices.value = current + ScannedDevice(device, name)
                _events.tryEmit("Found: ${name ?: "unknown"} [${device.address}]")
            }
        }

        override fun onScanFailed(errorCode: Int) {
            _events.tryEmit("Scan failed: error $errorCode")
            _state.value = State.Idle
        }
    }

    // ── Connection ───────────────────────────────────────────────────────

    fun connect(device: BluetoothDevice) {
        stopScan()
        _state.value = State.Connecting
        _events.tryEmit("Connecting to ${device.address} …")
        gatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
    }

    fun disconnect() {
        gatt?.disconnect()
        gatt?.close()
        gatt = null
        csChar = null
        scChar = null
        _state.value = State.Disconnected
        _events.tryEmit("Disconnected")
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    _events.tryEmit("Connected, discovering services …")
                    gatt.requestMtu(517) // Request max MTU for BLE 5.x
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    _state.value = State.Disconnected
                    _events.tryEmit("Disconnected (status=$status)")
                }
            }
        }

        override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
            _events.tryEmit("MTU negotiated: $mtu bytes")
            gatt.discoverServices()
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status != BluetoothGatt.GATT_SUCCESS) {
                _events.tryEmit("Service discovery failed: status=$status")
                return
            }

            val service = gatt.getService(NdnBleProtocol.SERVICE_UUID)
            if (service == null) {
                _events.tryEmit("NDN BLE service not found!")
                return
            }

            csChar = service.getCharacteristic(NdnBleProtocol.CS_CHAR_UUID)
            scChar = service.getCharacteristic(NdnBleProtocol.SC_CHAR_UUID)

            if (csChar == null || scChar == null) {
                _events.tryEmit("NDN characteristics not found (CS=${csChar != null}, SC=${scChar != null})")
                return
            }

            // Subscribe to SC notifications.
            gatt.setCharacteristicNotification(scChar!!, true)
            val descriptor = scChar!!.getDescriptor(
                java.util.UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
            )
            if (descriptor != null) {
                gatt.writeDescriptor(
                    descriptor,
                    BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                )
            }

            _state.value = State.Connected
            _events.tryEmit("Ready — CS and SC characteristics found, notifications enabled")
        }

        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray
        ) {
            if (characteristic.uuid == NdnBleProtocol.SC_CHAR_UUID) {
                _events.tryEmit("SC notification: ${value.size} bytes")
                _rxData.tryEmit(value)
            }
        }
    }

    // ── Write ────────────────────────────────────────────────────────────

    /** Write raw bytes to the CS (client→server) characteristic. */
    fun writeCs(data: ByteArray): Boolean {
        val cs = csChar ?: return false
        val g = gatt ?: return false
        val result = g.writeCharacteristic(
            cs,
            data,
            BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
        )
        return result == BluetoothStatusCodes.SUCCESS
    }
}

File: examples/ble/ios/BleTest/BleClient.swift
import CoreBluetooth
import Combine
import Foundation

/// BLE GATT central client for NDN-over-BLE testing.
///
/// Scans for the NDN BLE service, connects, and provides read/write access
/// to the CS (write) and SC (notify) characteristics.
class BleClient: NSObject, ObservableObject {
    enum State: Equatable {
        case idle
        case scanning
        case connecting
        case connected
        case disconnected
    }

    struct ScannedDevice: Identifiable {
        let id: UUID
        let peripheral: CBPeripheral
        let name: String
    }

    @Published var state: State = .idle
    @Published var scannedDevices: [ScannedDevice] = []
    @Published var log: [String] = []

    let rxData = PassthroughSubject<Data, Never>()

    private var centralManager: CBCentralManager!
    private var connectedPeripheral: CBPeripheral?
    private var csCharacteristic: CBCharacteristic?
    private var scCharacteristic: CBCharacteristic?

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    func appendLog(_ msg: String) {
        let ts = ISO8601DateFormatter().string(from: Date())
        DispatchQueue.main.async {
            self.log.append("[\(ts)] \(msg)")
        }
    }

    // MARK: - Scanning

    func startScan() {
        scannedDevices = []
        state = .scanning
        appendLog("Scanning for NDN BLE peripherals ...")
        centralManager.scanForPeripherals(
            withServices: [NdnBleProtocol.serviceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )
    }

    func stopScan() {
        centralManager.stopScan()
        if state == .scanning { state = .idle }
    }

    // MARK: - Connection

    func connect(device: ScannedDevice) {
        stopScan()
        state = .connecting
        appendLog("Connecting to \(device.name) ...")
        connectedPeripheral = device.peripheral
        centralManager.connect(device.peripheral, options: nil)
    }

    func disconnect() {
        if let p = connectedPeripheral {
            centralManager.cancelPeripheralConnection(p)
        }
        connectedPeripheral = nil
        csCharacteristic = nil
        scCharacteristic = nil
        state = .disconnected
        appendLog("Disconnected")
    }

    // MARK: - Write

    func writeCs(_ data: Data) -> Bool {
        guard let cs = csCharacteristic, let p = connectedPeripheral else { return false }
        p.writeValue(data, for: cs, type: .withoutResponse)
        return true
    }
}

// MARK: - CBCentralManagerDelegate

extension BleClient: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            appendLog("Bluetooth powered on")
        case .poweredOff:
            appendLog("Bluetooth powered off")
        case .unauthorized:
            appendLog("Bluetooth unauthorized")
        default:
            appendLog("Bluetooth state: \(central.state.rawValue)")
        }
    }

    func centralManager(_ central: CBCentralManager,
                         didDiscover peripheral: CBPeripheral,
                         advertisementData: [String: Any],
                         rssi RSSI: NSNumber) {
        let name = peripheral.name ?? advertisementData[CBAdvertisementDataLocalNameKey] as? String ?? "unknown"
        if !scannedDevices.contains(where: { $0.peripheral.identifier == peripheral.identifier }) {
            let device = ScannedDevice(id: peripheral.identifier, peripheral: peripheral, name: name)
            DispatchQueue.main.async {
                self.scannedDevices.append(device)
            }
            appendLog("Found: \(name) [\(peripheral.identifier)]")
        }
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        appendLog("Connected, discovering services ...")
        peripheral.delegate = self
        peripheral.discoverServices([NdnBleProtocol.serviceUUID])
    }

    func centralManager(_ central: CBCentralManager,
                         didFailToConnect peripheral: CBPeripheral,
                         error: Error?) {
        state = .disconnected
        appendLog("Connection failed: \(error?.localizedDescription ?? "unknown")")
    }

    func centralManager(_ central: CBCentralManager,
                         didDisconnectPeripheral peripheral: CBPeripheral,
                         error: Error?) {
        state = .disconnected
        appendLog("Disconnected: \(error?.localizedDescription ?? "clean")")
    }
}

// MARK: - CBPeripheralDelegate

extension BleClient: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil else {
            appendLog("Service discovery error: \(error!)")
            return
        }
        guard let service = peripheral.services?.first(where: { $0.uuid == NdnBleProtocol.serviceUUID }) else {
            appendLog("NDN BLE service not found!")
            return
        }
        peripheral.discoverCharacteristics(
            [NdnBleProtocol.csCharUUID, NdnBleProtocol.scCharUUID],
            for: service
        )
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didDiscoverCharacteristicsFor service: CBService,
                     error: Error?) {
        guard error == nil else {
            appendLog("Characteristic discovery error: \(error!)")
            return
        }
        for char in service.characteristics ?? [] {
            if char.uuid == NdnBleProtocol.csCharUUID {
                csCharacteristic = char
            } else if char.uuid == NdnBleProtocol.scCharUUID {
                scCharacteristic = char
                peripheral.setNotifyValue(true, for: char)
            }
        }

        if csCharacteristic != nil && scCharacteristic != nil {
            state = .connected
            appendLog("Ready -- CS and SC characteristics found, notifications enabled")
        } else {
            appendLog("Missing characteristics (CS=\(csCharacteristic != nil), SC=\(scCharacteristic != nil))")
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didUpdateValueFor characteristic: CBCharacteristic,
                     error: Error?) {
        guard error == nil, characteristic.uuid == NdnBleProtocol.scCharUUID,
              let value = characteristic.value else { return }
        appendLog("SC notification: \(value.count) bytes")
        rxData.send(value)
    }
}
